### ISSUE TITLE
Generalize RouteMismatch

### ISSUE THREAD

**Issue by jkarni** [haskell-servant/servant#48]
_From @jkarni on January 30, 2015 4:52_

https://github.com/haskell-servant/servant-server/pull/22 adds the ability to have some overriding custom errors in `RouteMismatch`. It'd be nice to generalize this to a more flexible priority system, as discussed in the PR.

_Copied from original issue: haskell-servant/servant-server#23_

**Comment by jkarni** [comment]
_From @christian-marie on January 30, 2015 23:26_

The most obvious and simple way appears (to me) to be just a typeclass, such that you can failWith these things if they can produce an appropriate HTTP response.

**Comment by jkarni** [comment]
_From @alpmestan on January 30, 2015 23:32_

The thing is, we have to be able to "compare" them. Some mismatches mean "stop right here, i don't care about whether other endpoints could match", whereas some aren't as strong and the routing can actually lead to a successful match later on with another endpoint, despite the fact than an earlier endpoint returned a mismatch.

**Comment by jkarni** [comment]
_From @christian-marie on January 31, 2015 0:0_

Right, surely this information can be encoded in the typeclass?

**Comment by jkarni** [comment]
_From @alpmestan on January 31, 2015 10:25_

Have you read the discussion from [this PR](https://github.com/haskell-servant/servant-server/pull/22) ? I'm not sure how we could go with a typeclass, especially considering we would have to look up the "priority" of other mismatches by going to the source code, so not very handy. If you have any idea/suggestion, we're all ears! :)

**Comment by jkarni** [comment]
_From @thsutton on February 1, 2015 8:11_

I wonder if a better solution might be a priority from a small set (as in, e.g., syslog) together with an HTTP status: the priority controls routing behavior (stop now vs backtrack, etc.) and the HTTP status is what gets sent to the client.

This allows HasServer instances to distinguish between "nothing to see here, move along" and "if this was meant for me, the client made mistake X" and "mayday, mayday, mayday: abort now" without having to overload the HTTP status codes with additional meaning or try to tell a nice story about they way you assign Rational or `Ord a` values consistently in your handlers.

**Comment by jkarni** [comment]
"Stop now" vs. "continue" is one axis (`succeedWith` vs `failWith`) ; `RouteMismatch` is another, and I don't think we should or need to intermingle them.

**Comment by jkarni** [comment]
_From @christian-marie on February 1, 2015 19:48_

My intention with the typeclass was that it would have a priority field (ugly, but we've all seen it before).

```haskell

class Errorablething e where
    priority :: e -> Int/Float
    response :: e -> ResponseStuff

data ESorting = forall e. Errorablething e => ESorting e

instance Ord ESorting where
    compare = comparing priority

tehError :: [ESorting] -> ESorting
tehError = maximum
```

That kind of thing.

I've had a bit of a think about the over-engineered alternatives (along the lines of datatypes al-a-carte) and they either won't work or aren't worth it in my opinion.

**Comment by jkarni** [comment]
I don't get the advantage of a typeclass over just a datatype. Could you explain?

**Comment by jkarni** [comment]
_From @christian-marie on February 2, 2015 2:34_

Yes, the type class is  just over-complicating things now that I consider it further.

**Comment by jkarni** [comment]
Probably should be included with the changes kosmikus suggested [here](https://gist.github.com/kosmikus/79310bc1e40f6d651a0d)

**Comment by jkarni** [comment]
Closed in #218
